internal/cmd/envoy/shutdown_manager.go

```diff
@@ -0,0 +1,209 @@
+)
+
+var (
+	logger = logging.DefaultLogger(v1alpha1.LogLevelInfo).WithName("shutdown-manager")
+)
+
+const (
+	// ShutdownReadyPort is the port Envoy shutdown manager will listen on.
+	ShutdownReadyPort = 19002
+	// ShutdownReadyFile is the file used to indicate shutdown readiness.
+	ShutdownReadyFile = "/tmp/shutdown-ready"
+)
+
+// ShutdownManager serves shutdown manager process for Envoy proxies.
+func ShutdownManager(readyTimeout time.Duration) error {
+	// Setup HTTP handler
+	handler := http.NewServeMux()
+	handler.HandleFunc("/shutdown/ready", func(w http.ResponseWriter, r *http.Request) {
+		shutdownReadyHandler(w, r, ShutdownReadyFile)
+	})
+
+	// Setup HTTP server
+	srv := http.Server{
+		Handler:           http.TimeoutHandler(handler, readyTimeout, ""),
+		Addr:              fmt.Sprintf(":%d", ShutdownReadyPort),
+		ReadTimeout:       5 * time.Second,
+		ReadHeaderTimeout: 5 * time.Second,
+		WriteTimeout:      10 * time.Second,
+		IdleTimeout:       15 * time.Second,
+	}
+
+	// Setup signal handling
+	c := make(chan struct{})
+	go func() {
+		s := make(chan os.Signal, 1)
+		signal.Notify(s, os.Interrupt, syscall.SIGTERM)
+
+		r := <-s
+		logger.Info(fmt.Sprintf("received %s", unix.SignalName(r.(syscall.Signal))))
+
+		// Shutdown HTTP server without interrupting active connections
+		if err := srv.Shutdown(context.Background()); err != nil {
+			logger.Error(err, "server shutdown error")
+		}
+		close(c)
+	}()
+
+	// Start HTTP server
+	logger.Info("starting shutdown manager")
+	if err := srv.ListenAndServe(); !errors.Is(err, http.ErrServerClosed) {
+		logger.Error(err, "starting shutdown manager failed")
+	}
+
+	// Wait until done
+	<-c
+	return nil
+}
+
+// shutdownReadyHandler handles the endpoint used by a preStop hook on the Envoy
+// container to block until ready to terminate. After the graceful drain process
+// has completed a file will be written to indicate shutdown readiness.
+func shutdownReadyHandler(w http.ResponseWriter, _ *http.Request, readyFile string) {
+	logger.Info("received shutdown ready request")
+
+	// Poll for shutdown readiness
+	for {
+		_, err := os.Stat(readyFile)
+		switch {
+		case os.IsNotExist(err):
+			time.Sleep(1 * time.Second)
+		case err != nil:
+			logger.Error(err, "error checking for shutdown readiness")
+		case err == nil:
+			logger.Info("shutdown readiness detected")
+			w.WriteHeader(http.StatusOK)
+			return
+		}
+	}
+}
+
+// Shutdown is called from a preStop hook where it will block until envoy can
+// gracefully drain open connections prior to pod shutdown.
+func Shutdown(drainTimeout time.Duration, minDrainDuration time.Duration, exitAtConnections int) error {
+	var startTime = time.Now()
+	var allowedToExit = false
+
+	// Reconfigure logger to write to stdout of main process if running in Kubernetes
+	if _, k8s := os.LookupEnv("KUBERNETES_SERVICE_HOST"); k8s && os.Getpid() != 1 {
+		logger = logging.FileLogger("/proc/1/fd/1", "shutdown-manager", v1alpha1.LogLevelInfo)
```

Comment:
There's a discussion [here](https://www.reddit.com/r/docker/comments/97src6/application_output_to_docker_logs_works_only_when/.) about this option maybe not working if running the process as non-root. 
Not sure if this concern is applicable in this case...

Reply:
> writing directly to /proc/1/fd/1 will be problematic if the process drops root privileges but PID 1 is still running as root

I don't think it will apply in this case, as neither PID 1 nor the exec'd process are running as root since the container sets the uid:gid in the build ([see here](https://github.com/envoyproxy/gateway/blob/main/tools/docker/envoy-gateway/Dockerfile#L7))